route/time_dependent.go

```diff
@@ -0,0 +1,150 @@
+package route
+
+import (
+	"errors"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/nextmv-io/sdk/measure"
+	"github.com/nextmv-io/sdk/model"
+)
+
+// ByIndexAndTime holds a Measure and an EndTime (exclusive) up until this
+// measure is to be used as a Unix timestamp. ByIndexAndTime is to be used with
+// NewTimeDependentMeasure which a slice of ByIndexAndTime.
+type ByIndexAndTime struct {
+	Measure ByIndex
+	EndTime int
+}
+
+// ClientOption can pass options to be used with a TimeDependentMeasure client.
+type ClientOption func(*client)
+
+// TimeDependentMeasuresClient is an interface to handle time dependent
+// measures. It implements a Cost function that takes time into account to
+// calculate costs.
+type TimeDependentMeasuresClient interface {
+	Cost() func(from, to int, data measure.VehicleData) float64
+	DependentByIndex() measure.DependentByIndex
+}
+
+type client struct {
+	measures        []ByIndexAndTime
+	fallbackMeasure ByIndexAndTime
+	cache           sync.Map
```

Comment:
What motivated you to introduce a cache vs. maintaining a static mapping from time to measure?

Reply:
basically a full pre-cache was a wish @davidrijsman had from past experiences. to leave it open how to use cache, the full pre-cache is an option that can be passed. otherwise the mapping will be created on the fly and stored for future usages.

Reply:
My gut says there is no need for a cache. If I understand it correctly, given a startTime , we will choose the first measure whose endTime is greater than startTime.

We know that the runtime overhead is always constant in the number of stops/vehicles and only depends on the number of measures.

For a smaller number of measures we can simply loop through the sorted list of measures and return as soon as the above relation is true: `O(n), where n is the number of measures`, but very simple code. Since the list is sorted, we can also use a binary search that brings us to `O(log n)` run time probably. The overhead in code complexity makes only sense if we expect a very large number of measures.

I have the feeling these two approaches would outperform the cache based approach, both in terms of runtime and code complexity. Only for very large inputs I would expect the cache to perform better. But haven't tested it - maybe I am totally wrong here.

Reply:
If we don't need a cache, then I think the logic to create a time dependent measure can be greatly simplified I think.

Reply:
Until we have some good benchmarks setup for this case, we could go with the O(1) approach under the hood.

Reply:
done